core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -4882,4 +4947,282 @@ private boolean isH2DB(Connection dbConnection) throws Exception {
+     */
+    private Connection getConnectionOnCircuitBreakerClose() throws CircuitBreakerException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Connection circuit breaker state: " + getCircuitBreakerState()
+                    + ", so trying to obtain the connection for domain " + getMyDomainName());
+        }
+        int retryCounter = 0;
+        while (true) {
+            try {
+                return getConnection();
+            } catch (Exception e) {
+                log.warn("Error occurred while obtaining JDBC connection for domain " + getMyDomainName()
+                + ". Hence, retry attempt to recover DB connection: " + retryCounter);
+
+                if (++retryCounter >= getConnectionRetryCount()) {
+                    log.error("Retry count exceeds above the maximum count: " + getConnectionRetryCount() +
+                            " and failed for domain " + getMyDomainName());
+                    setCircuitBreakerState(CIRCUIT_STATE_OPEN);
+                    setThresholdStartTime(System.currentTimeMillis());
+                    throw new CircuitBreakerException("Error occurred while obtaining connection for domain: "
+                            + getMyDomainName() + " and circuit breaker state set to: " + getCircuitBreakerState(), e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Sets threadsafe JDBC ConnectionCircuitBreakerState
+     *
+     * @param jdbcConnectionCircuitBreakerState JDBC ConnectionCircuitBreakerState
+     */
+    public void setCircuitBreakerState(String jdbcConnectionCircuitBreakerState) {
+
+        writeLock.lock();
+        try{
+            this.jdbcConnectionCircuitBreakerState = jdbcConnectionCircuitBreakerState;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Sets threadsafe thresholdTimeoutInMilliseconds
+     *
+     * @param thresholdTimeoutInMilliseconds threshold Timeout in Milliseconds
+     */
+    public void setThresholdTimeoutInMilliseconds(long thresholdTimeoutInMilliseconds) {
+
+        writeLock.lock();
+        try{
+            this.thresholdTimeoutInMilliseconds = thresholdTimeoutInMilliseconds;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Sets threadsafe thresholdStartTime
+     *
+     * @param thresholdStartTime Threshold Start Time
+     */
+    public void setThresholdStartTime(long thresholdStartTime) {
+
+        writeLock.lock();
+        try{
+            this.thresholdStartTime = thresholdStartTime;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Sets threadsafe connectionRetryCount
+     *
+     * @param connectionRetryCount Connection Retry Count
+     */
+    public void setConnectionRetryCount(int connectionRetryCount) {
+
+        writeLock.lock();
+        try{
+            this.connectionRetryCount = connectionRetryCount;
```

Comment:
isn't setConnectionRetryCount a configuration values ?

then we don't need to write lock.

Reply:
yes, it is. let's remove the write lock for this config property.